### PR TITLE
Add benchmarks for `Gen.sample`

### PR DESCRIPTION
**Pull request by adam-becker**
This PR puts the examples from #135 into a proper benchmark, in support of #371.

/cc @moodmosaic @TysonMN

### CODE REVIEW

**Comment by adam-becker** [comment]
``` ini

BenchmarkDotNet=v0.13.1, OS=debian 10 (container)
Intel Core i7-9750H CPU 2.60GHz, 8 CPU, 8 logical and 8 physical cores
.NET Core SDK=3.1.413
  [Host]        : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT DEBUG
  .NET Core 3.1 : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT

Job=.NET Core 3.1  Runtime=.NET Core 3.1  

```
|         Method |     Mean |   Error |  StdDev |
|--------------- |---------:|--------:|--------:|
| GenSampleLarge | 131.1 ms | 1.57 ms | 1.39 ms |

Should I add FsCheck as well to compare against?

**Comment by adam-becker** [comment]
Also, as a fun side note I removed `bigint` again (again, naively just replacing `bigint` with `int64`) and got the following:

``` ini

BenchmarkDotNet=v0.13.1, OS=debian 10 (container)
Intel Core i7-9750H CPU 2.60GHz, 8 CPU, 8 logical and 8 physical cores
.NET Core SDK=3.1.413
  [Host]        : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT DEBUG
  .NET Core 3.1 : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT

Job=.NET Core 3.1  Runtime=.NET Core 3.1  

```
|                 Method |     Mean |    Error |   StdDev |
|----------------------- |---------:|---------:|---------:|
| GenSampleLargeHedgehog | 52.42 ms | 0.695 ms | 0.616 ms |
|  GenSampleLargeFsCheck | 42.26 ms | 0.352 ms | 0.330 ms |

File: tests/Hedgehog.Benchmarks/GenBenchmarks.fs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        |> HGen.sample 100 100000
+        |> Seq.iter ignore
+
+    [<Benchmark>]
+    member __.GenSampleLargeFsCheck () =
+        ['0'..'9'] @ ['A'..'Z'] @ ['a'..'z']
+        |> FGen.elements
+        |> FGen.listOfLength 10
+        |> FGen.map (List.toArray >> System.String.Concat)
+        |> FGen.sample 0 10000
```
Comment (TysonMN) [inline tests/Hedgehog.Benchmarks/GenBenchmarks.fs:26]: This looks like you are comparing apples to oranges.  Most surprisingly, it looks to me like you are sampling from the FsCheck generator 10,000 times but sampling from the Hedgehog generator 100,000 times.

Can you add to the method names what is being sample and how many times?

Comment (adam-becker) [reply tests/Hedgehog.Benchmarks/GenBenchmarks.fs:26]: 🤦🏼‍♂️ Good catch! I got my wires crossed with the two examples in #135, this has been fixed. If anything, it gives us even more room for improvement 😅 

``` ini
BenchmarkDotNet=v0.13.1, OS=debian 10 (container)
Intel Core i7-4790K CPU 4.00GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.409
  [Host]        : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT DEBUG
  .NET Core 3.1 : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT

Job=.NET Core 3.1  Runtime=.NET Core 3.1  
```
|                 Method |      Mean |    Error |   StdDev |
|----------------------- |----------:|---------:|---------:|
| GenSampleLargeHedgehog | 125.95 ms | 0.336 ms | 0.298 ms |
|  GenSampleLargeFsCheck |  26.18 ms | 0.237 ms | 0.210 ms |

**Comment by moodmosaic** [comment]
I need to refresh my memory on the choice for `bigint`, perhaps because of the stuff in `Numeric.fs` 🤔

**Comment by TysonMN** [comment]
Add `Params` like in [this file](https://github.com/TysonMN/Elmish.WPF/blob/benchmark/merge/src/Elmish.WPF.Benchmarks/Program.fs).  I am thinking powers of 10 would be good choices.

**Comment by adam-becker** [comment]
> I need to refresh my memory on the choice for `bigint`, perhaps because of the stuff in `Numeric.fs` 🤔

I think it was out of concern for generating min/max ranges for `uint64`/`int64`? I remember seeing a comment from you on some issue or PR about it. Either way, I think we can safely replace it 🤷🏼‍♂️

**Comment by adam-becker** [comment]
> Add `Params` like in [this file](https://github.com/TysonMN/Elmish.WPF/blob/benchmark/merge/src/Elmish.WPF.Benchmarks/Program.fs). I am thinking powers of 10 would be good choices.
> 
> And this slightly contradicts my other comment that said to put the quantity to be sampled in the method name. With this change, don't put the quantity in the name.

``` ini

BenchmarkDotNet=v0.13.1, OS=debian 10 (container)
Intel Core i7-9750H CPU 2.60GHz, 8 CPU, 8 logical and 8 physical cores
.NET Core SDK=3.1.413
  [Host]        : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT DEBUG
  .NET Core 3.1 : .NET Core 3.1.19 (CoreCLR 4.700.21.41101, CoreFX 4.700.21.41603), X64 RyuJIT

Job=.NET Core 3.1  Runtime=.NET Core 3.1

```
|                 Method |      N |          Mean |        Error |       StdDev |        Median |
|----------------------- |------- |--------------:|-------------:|-------------:|--------------:|
| **HedgehogGenSampleInt32** |    **100** |     **139.42 μs** |     **1.612 μs** |     **1.508 μs** |     **139.31 μs** |
|  FsCheckGenSampleInt32 |    100 |      23.39 μs |     0.065 μs |     0.054 μs |      23.38 μs |
| **HedgehogGenSampleInt32** |   **1000** |   **1,313.18 μs** |    **17.493 μs** |    **16.363 μs** |   **1,304.07 μs** |
|  FsCheckGenSampleInt32 |   1000 |     169.81 μs |     3.191 μs |     4.260 μs |     170.33 μs |
| **HedgehogGenSampleInt32** |  **10000** |  **13,045.48 μs** |   **185.140 μs** |   **173.180 μs** |  **13,081.98 μs** |
|  FsCheckGenSampleInt32 |  10000 |   1,878.52 μs |    31.893 μs |    29.833 μs |   1,886.14 μs |
| **HedgehogGenSampleInt32** | **100000** | **130,364.13 μs** | **1,713.186 μs** | **1,602.515 μs** | **130,159.68 μs** |
|  FsCheckGenSampleInt32 | 100000 |  27,397.14 μs |   542.102 μs | 1,309.237 μs |  26,793.29 μs |

**Comment by TysonMN** [comment]
This looks great. Excellent work @adam-becker!

**Review by TysonMN — APPROVED** [review]
I think this PR is ready to merge

**Comment by dharmaturtle** [comment]
I don't have any experience with BenchmarkDotNet - is it possible to use it to write a test that will fail if, say, HedgehogGenSampleInt32 N=100 takes 1 second? Not a blocker to merging - I'm just wondering.

**Comment by TysonMN** [comment]
I don't think so.  The point of a BenchmarkDotNet test is to look at the report it generates.

**Pull request merged**